server/src/main/java/org/candlepin/dto/api/v1/PoolDTO.java

```diff
@@ -0,0 +1,1904 @@
+import com.fasterxml.jackson.annotation.JsonInclude;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
+import com.fasterxml.jackson.databind.annotation.JsonSerialize;
+import io.swagger.annotations.ApiModel;
+
+import org.apache.commons.lang.builder.EqualsBuilder;
+import org.apache.commons.lang.builder.HashCodeBuilder;
+import org.candlepin.common.jackson.HateoasInclude;
+import org.candlepin.jackson.CandlepinAttributeDeserializer;
+import org.candlepin.jackson.CandlepinLegacyAttributeSerializer;
+import org.candlepin.model.Consumer;
+import org.candlepin.util.MapView;
+import org.candlepin.util.SetView;
+
+import javax.xml.bind.annotation.XmlAccessType;
+import javax.xml.bind.annotation.XmlAccessorType;
+import javax.xml.bind.annotation.XmlRootElement;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * A DTO representation of the Pool entity
+ */
+@XmlRootElement(name = "pool")
+@XmlAccessorType(XmlAccessType.PROPERTY)
+@ApiModel(parent = TimestampedCandlepinDTO.class, description = "DTO representing a Pool")
+@JsonFilter("PoolFilter")
+public class PoolDTO extends TimestampedCandlepinDTO<PoolDTO> implements  LinkableDTO {
+    public static final long serialVersionUID = 1L;
+
+    /**
+     * Internal DTO object for Product
+     */
+    public static class ProductDTO {
+        private final String id;
+        private final String name;
+
+        @JsonCreator
+        public ProductDTO(
+            @JsonProperty("id") String id,
+            @JsonProperty("name") String name) {
+            if (id == null || id.isEmpty()) {
+                throw new IllegalArgumentException("The product id is null or empty.");
+            }
+
+            this.id = id;
+            this.name = name;
+        }
+
+        public String getId() {
+            return this.id;
+        }
+
+        public String getName() {
+            return this.name;
+        }
+
+        @Override
+        public boolean equals(Object obj) {
+            if (obj == this) {
+                return true;
+            }
+
+            if (obj instanceof ProductDTO) {
+                ProductDTO that = (ProductDTO) obj;
+
+                EqualsBuilder builder = new EqualsBuilder()
+                    .append(this.getId(), that.getId())
+                    .append(this.getName(), that.getName());
+
+                return builder.isEquals();
+            }
+
+            return false;
+        }
+
+        @Override
+        public int hashCode() {
+            HashCodeBuilder builder = new HashCodeBuilder(37, 7)
+                .append(this.getId())
+                .append(this.getName());
+
+            return builder.toHashCode();
+        }
+    }
+
+    /**
+     * Internal DTO object for ProvidedProduct
+     */
+    @JsonFilter("ProvidedProductFilter")
+    public static class ProvidedProductDTO {
```

Comment:
They're not identical to the client though. From what I see on the spec test candlepin_api.rb, the id field on each of these fields has a different name. For the product/derivedProduct, it is called 'id', while for the providedProducts/derivedProvidedProducts collections, it is called 'productId'. Also, I cannot amend this behaviour by having a single internal ProductDTO class and using two constructors because adding two @JsonCreator annotations is not supported.

Reply:
@Ceiu The question I wanted to ask is about this^. Is keeping the two identical internal DTOs fine, since I don't see an alternative solution?